README.md
# ESLint Configurations

[![npm package](https://img.shields.io/npm/v/@sudaraka/eslint-config-default.svg)](https://www.npmjs.com/package/@sudaraka/eslint-config-default)

Shared [ESLint](http://eslint.org/) configurations used for my projects.

## License

[![BSD 2-Clause License](https://img.shields.io/badge/license-BSD%202--Clause-blue.svg)](http://opensource.org/licenses/BSD-2-Clause)

Copyright 2015-2018 Sudaraka Wijesinghe <sudaraka@sudaraka.org>

This program comes with ABSOLUTELY NO WARRANTY;
This is free software, and you are welcome to redistribute it and/or modify it
under the terms of the BSD 2-clause License. See the LICENSE file for more
details.


CHANGELOG.md
# CHANGE LOG

All notable changes to the eslint-config-default project.

Copyright 2015-2018 Sudaraka Wijesinghe <sudaraka@sudaraka.org>

This program comes with ABSOLUTELY NO WARRANTY;
This is free software, and you are welcome to redistribute it and/or modify it
under the terms of the BSD 2-clause License. See the LICENSE file for more
details.

---

## [2.5.0] - 2018-02-08
### Added
- All published rules as of `eslint-plugin-react` 7.6.1

### Changed
- Update ESLint dependency to 4.17.0 (no rules changes).
- Disabled `multiline-comment-style`, all available options conflict with my
  coding style.

## [2.4.0] - 2017-11-30
### Changed
- Disabled `no-inline-comments`, want to use inline comments to hint argument
  names, types, etc.

## [2.3.0] - 2017-11-29
### Added
- All published rules as of ESLint 4.12.0
- All published rules as of `eslint-plugin-react` 7.5.1
  - `jsx-curly-brace-presence` off by default to allow use of braces only when
    necessary.

### Changed
- Set `ESLint` as a peer dependency.

## [2.2.0] - 2017-09-02
### Added
- All published rules as of ESLint 4.6.0
- Graceful handling of remote data errors.

### Changes
- Turn off rule package version tag processing by default. *(version tag processing can be enabled by command line parameter)*
- Allow modification of event properties (i.e. `e.target.disabled = true`)
- Allow magic numbers 0, 1 and 2.
  > 0, 1 allowed for length/emptiness test (i.e. `1 > length` or `0 < length`)
  >
  > 2 allowed for decimal point representation (i.e. `value.toFixed(2)`)
- Adjust for `unpkg.com` URL change. (`?json` => `?meta`)
- Removed some clutter using argument spread.

### Fixed
- Prevent errors from being swallowed at `Promise.catch` handlers in
  mid-process.

## [2.1.0] - 2017-08-20
### Changed
- Set default behavior for following rules to `error`:
  - `array-callback-return`
  - `consistent-return`
  - `no-empty-function`
  - `no-magic-numbers` (with `ignoreArrayIndexes: true`)
  - `no-param-reassign` (with `props: true`)
  - `camelcase`
  - `no-inline-comments`
  - `padded-blocks` (never)
  - `init-declarations` (always)

## [2.0.0] - 2017-08-19
### Added
- `.eslintrc` and `npm` script command to self-lint file in this package.

### Changed
- Switch to use `line-aligned` for JSX closing brackets.
- Turn off all deprecated rules as of ESLint 4.5.0
- Split default rule set into multiple modules.

### Fixed
- Lint error in .js files.

## [1.10.0] - 2017-08-10
### Added
- All published rules as of `eslint-plugin-react` 7.2.0

## [1.9.0] - 2017-07-09
### Added
- All published rules as of ESLint 4.2.0
- All published rules as of `eslint-plugin-react` 7.1.0

## [1.8.0] - 2017-06-12
### Added
- All published rules as of ESLint 4.0.0
- All published rules as of `eslint-plugin-react` 7.0.1
- checkrules.js: lookup multiple release tags in same package.

### Removed
- Deprecated rules from `react` plugin.

### Changed
- Adjust `no-multi-spaces` rule to allow multiple (conventionally 2) spaced
  before EOL comment.

## [1.7.0] - 2017-03-05
### Added
- All published rules as of ESLint 3.17.0

## [1.6.0] - 2017-02-26
### Added
- checkrules.js: introduce optional rule name prefixes to handle internal
  rule name vs filename incompatibilities in plugins.
- All published rules as of `eslint-plugin-react` 6.10.0

### Changed
- checkrules.js: Rename "prefix" field in source dataset to "namespace".
- checkrules.js: Adjust for ESLint v3.16 package structure.

### Fixed
- `prefixes` field of `RULE_SOURCES` always have to be an Array.

## [1.5.0] - 2017-02-04
### Added
- Extended configuration file for [Preact](https://preactjs.com/) based projects.
- All published rules as of ESLint 3.15.0

## [1.4.0] - 2017-01-21
### Added
- All published rules as of ESLint 3.14.0
- Source to check released rules of `eslint-react-plugin` via unpkg.com.

### Changed
- Get package information from `npm` repository.
- Set `ESLint` as direct dependency.

### Removed
- Sources for unreleased rules on github.

## [1.3.0] - 2017-01-20
### Added
- Support to process rules from multiple locations in`checkrules.js`.
- Optional/Custom function to process rules data retrieved from the remote
  source.
- Support rule prefixes.
- Source to check rules of `eslint-react-plugin`.
- Available rules of `eslint-plugin-react` on 2017-01-20.

### Changed
- Use `chalk` for colorful display of the result.

## [1.2.0] - 2017-01-07
### Added
- All published rules as of ESLint 3.13.0

## [1.1.1] - 2016-12-11
### Changed
- Turned off `capitalized-comments` rule.

## [1.1.0] - 2016-12-09
### Added
- All published rules as of ESLint 3.12.0
- All published rules as of `eslint-plugin-react` 6.4.1

## [1.0.0] - 2016-10-15
### Added
- All published rules as of ESLint 3.8.0

### Changed
- Allow _ to indicate unused argument.
- Do not apply `max-len` rule to strings and template literals.

## [0.7.3] - 2016-09-17
### Changed
- Make `max-len` for code 100 to allow some flexibility. 80 should be the limit
  by discipline/practice.

## [0.7.2] - 2016-09-10
### Fixed
- Invalid value in max-line rule ('error' => 2)

## [0.7.1] - 2016-09-10
### Added
- All published rules as of ESLint 3.5.0

### Changed
- Rename npmcdn.com to unpkg.com (via https://github.com/npmcdn-to-unpkg-bot PR,
  Thanks).

### Fixed
- Removed some duplicated entries in default rules.

## [0.7.0] - 2016-08-13
### Added
- All published rules as of ESLint 3.3.0
- All published rules as of `eslint-plugin-react` 6.0.0

### Changed
- Allow underscore as unused variable for throw aways.
- checkrules.js: show different shape icons for released and beta rules.
- checkrules.js: show github link for beta rules.
- checkrules.js: get latest `ESLint` version from `npm` repository instead of
  from `package.json`.
- checkrules.js: Remove unwanted line breaks from output.

## [0.6.0] - 2016-07-16
### Added
- All published rules as of ESLint 3.1.0

### Changed
- Adjusted strict mode setting for ES5 configuration.

rules/stylistic-issues-a2m.js
/**
 * rules/stylistic-issues-a2m.js: rules listed as Stylistic Issues in ESLint docs
 *
 * Copyright 2017 Sudaraka Wijesinghe <sudaraka@sudaraka.org>
 *
 * This program comes with ABSOLUTELY NO WARRANTY;
 * This is free software, and you are welcome to redistribute it and/or modify
 * it under the terms of the BSD 2-clause License. See the LICENSE file for more
 * details.
 *
 */

/*
 * NOTE: this file only contain rules starting from A to M
 * See https://eslint.org/docs/rules/#stylistic-issues
 *
 */

module.exports = {
  'array-bracket-newline': [
    'error',
    { 'minItems': 2 }
  ],
  'array-bracket-spacing': [
    'error',
    'always'
  ],
  'array-element-newline': [
    'error',
    { 'minItems': 2 }
  ],
  'block-spacing': [
    'error',
    'always'
  ],
  'brace-style': [
    'error',
    'stroustrup'
  ],
  'camelcase': 'error',
  'capitalized-comments': 'off',
  'comma-dangle': [
    'error',
    'never'
  ],
  'comma-spacing': [
    'error',
    {
      'before': false,
      'after': true
    }
  ],
  'comma-style': [
    'error',
    'last'
  ],
  'computed-property-spacing': [
    'error',
    'never'
  ],
  'consistent-this': [
    'error',
    'self'
  ],
  'eol-last': 'error',
  'func-call-spacing': [
    'error',
    'never'
  ],
  'func-name-matching': 'error',
  'func-names': 'off',
  'func-style': [
    'error',
    'expression'
  ],
  'function-paren-newline': [
    'error',
    'consistent'
  ],
  'id-blacklist': 'off',
  'id-length': 'off',
  'id-match': [
    'error',
    '^[a-zA-Z0-9_$]+([a-z0-9_]+)*$',
    { 'properties': true }
  ],
  'implicit-arrow-linebreak': [
    'error',
    'beside'
  ],
  'indent': [
    'error',
    2
  ],
  'indent-legacy': 'off',
  'jsx-quotes': [
    'error',
    'prefer-single'
  ],
  'key-spacing': [
    'error',
    {
      'beforeColon': false,
      'afterColon': true
    }
  ],
  'keyword-spacing': [
    'error',
    {
      'after': false,
      'before': true,
      'overrides': {
        'else': { 'after': true },
        'export': { 'after': true },
        'from': { 'after': true },
        'import': { 'after': true },
        'return': { 'after': true },
        'try': { 'after': true }
      }
    }
  ],
  'line-comment-position': [
    'error',
    { 'position': 'above' }
  ],
  'linebreak-style': [
    'error',
    'unix'
  ],
  'lines-around-comment': [
    'error',
    {
      'allowBlockStart': true,
      'beforeBlockComment': true,
      'beforeLineComment': false
    }
  ],
  'lines-between-class-members': [
    'error',
    'always'
  ],
  'lines-around-directive': 'off',
  'max-depth': [
    'error',
    8
  ],
  'max-len': [
    'error',
    {
      // 80 should be soft limit for code, using 100 here allow some flexibility
      'code': 100,
      'comments': 120,
      'ignoreComments': true,
      'ignoreTrailingComments': true,
      'ignoreUrls': true,
      'ignoreStrings': true,
      'ignoreTemplateLiterals': true,
      'tabWidth': 2
    }
  ],
  'max-lines': 'error',
  'max-nested-callbacks': [
    'error',
    5
  ],
  'max-params': [
    'error',
    5
  ],
  'max-statements': [
    'error',
    15
  ],
  'max-statements-per-line': [
    'error',
    { 'max': 1 }
  ],
  'multiline-comment-style': 'off',
  'multiline-ternary': 'off'
}
